src/linear_algebra/projection.lean

```diff
@@ -167,6 +167,87 @@ lemma linear_proj_of_is_compl_idempotent (h : is_compl p q) (x : E) :
     linear_proj_of_is_compl p q h x :=
 linear_proj_of_is_compl_apply_left h _
 
+lemma exists_unique_add_of_is_compl_prod (hc : is_compl p q) (x : E) :
+  ∃! (u : p × q), (u.fst : E) + u.snd = x :=
+(prod_equiv_of_is_compl _ _ hc).to_equiv.bijective.exists_unique _
+
+lemma exists_unique_add_of_is_compl (hc : is_compl p q) (x : E) :
+  ∃ (u : p) (v : q), ((u : E) + v = x ∧ ∀ (r : p) (s : q),
+    (r : E) + s = x → r = u ∧ s = v) :=
+let ⟨u, hu₁, hu₂⟩ := exists_unique_add_of_is_compl_prod hc x in
+  ⟨u.1, u.2, hu₁, λ r s hrs, prod.eq_iff_fst_eq_snd_eq.1 (hu₂ ⟨r, s⟩ hrs)⟩
+
+/-- Given linear maps `φ` and `ψ` from complement submodules, `of_is_compl` is
+the induced linear map over the entire module. -/
+def of_is_compl (h : is_compl p q) (φ : p →ₗ[R] F) (ψ : q →ₗ[R] F) : E →ₗ[R] F :=
+(linear_map.coprod φ ψ).comp (submodule.prod_equiv_of_is_compl _ _ h).symm
+
+@[simp] lemma of_is_compl_left_apply
+  (h : is_compl p q) {φ : p →ₗ[R] F} {ψ : q →ₗ[R] F} (u : p) :
+  of_is_compl h φ ψ (u : E) = φ u := by simp [of_is_compl]
+
+@[simp] lemma of_is_compl_right_apply
+  (h : is_compl p q) {φ : p →ₗ[R] F} {ψ : q →ₗ[R] F} (v : q) :
+  of_is_compl h φ ψ (v : E) = ψ v := by simp [of_is_compl]
+
+@[simp] lemma of_is_compl_zero (h : is_compl p q) :
+  (of_is_compl h 0 0 : E →ₗ[R] F) = 0 :=
+begin
+  ext, obtain ⟨_, _, rfl, _⟩ := exists_unique_add_of_is_compl h x,
+  simp [of_is_compl]
+end
```

Comment:
These proofs all look very similar. I think we can add the following lemma to turn them into one-liners:
```lean

lemma of_is_compl_eq (h : is_compl p q) {φ : p →ₗ[R] F} {ψ : q →ₗ[R] F} {χ : E →ₗ[R] F}
  (hφ : ∀ (v : p), φ v = χ v) (hψ : ∀ (v : q), ψ v = χ v) :
  of_is_compl h φ ψ = χ :=
begin
  ext x,
  obtain ⟨_, _, rfl, _⟩ := exists_unique_add_of_is_compl h x,
  simp [of_is_compl, hφ, hψ]
end
```

Reply:
I get the feeling that we don't need `exists_unique_add_of_is_compl` here at all, and can use something like `prod_whatever_equiv.injective` to build these proofs.

Reply:
I think `of_is_compl_eq` is a nice lemma to have regardless of what proof we use so I will add it.

Reply:
Another way of stateing those assumptions would be:
```
(hφ : φ = χ.comp p.subtype) (hψ : ψ = χ.comp q.subtype) 
```
which can be easier to prove sometimes using `ext`

Reply:
I've now added both versions